src/main/java/duke/storage/Storage.java

```diff
@@ -0,0 +1,179 @@
+     *
+     * @param filePath The location of the text file for tasks.
+     */
+    public Storage(String filePath) {
+        String[] pathSplitter = storageClassPath.split("/");
+        for (String directory: pathSplitter) {
+            if (!directory.isEmpty() && !directory.equals("build")) {
+                this.filePath += directory + "/";
+            } else if (directory.equals("build")) {
+                break;
+            }
+        }
+        this.filePath += filePath;
+    }
+
+    /**
+     * Updates the task list from reading the contents of the text file.
+     *
+     * @return ArrayList to update the task list.
+     * @throws IOException  If there is an error reading the text file.
+     */
+    public ArrayList<Task> read() throws IOException {
+        ArrayList<Task> items = new ArrayList<>();
+        Ui ui = new Ui();
+        File file = new File(filePath);
+
+        BufferedReader br = new BufferedReader(new FileReader(file));
+
+        String st;
+        String taskDesc;
+        String dateDesc;
+        String afterDesc;
+        String durDesc;
+        while ((st = br.readLine()) != null) {
+            String[] commandList = st.split("\\|");
+            try {
+                //clear previous dates/desc
+                taskDesc = "";
+                dateDesc = "";
+                afterDesc = "";
+                durDesc = "";
+                for (int i = ZERO; i < commandList.length; i++) {
+                    if (i == TWO) {
+                        taskDesc = commandList[i];
+                    } else if (i == THREE) {
+                        if (commandList[ZERO].equals("A")) {
+                            afterDesc = commandList[i];
+                        } else if (commandList[ZERO].equals("F")) {
+                            durDesc = commandList[i];
+                        } else {
+                            dateDesc = commandList[i];
+                        }
+                    }
+                }
+                boolean checked = false;
+                if (commandList.length > ONE) {
+                    if (!(commandList[ONE].equals("1") || commandList[ONE].equals("0"))) {
+                        throw new DukeException("Error reading 1 or 0, skipping to next line");
+                    }
+                    checked = commandList[ONE].equals("1");
+                }
+                Task t;
+                if (commandList[ZERO].equals("T")) {
+                    if (taskDesc.trim().isEmpty()) {
+                        throw new DukeException("Error reading description, skipping to next line");
+                    } else {
+                        t = new Todo(taskDesc);
+                        t.setStatusIcon(checked);
+                        items.add(t);
+                    }
+                } else if (commandList[ZERO].equals("D")) {
+                    if (taskDesc.trim().isEmpty() || dateDesc.trim().isEmpty()) {
```

Comment:
So much of repeated code. Think of an effective way of reducing it by moving the repeated validation check to single place